src/playground-components/weather-element.html

```diff
@@ -0,0 +1,55 @@
+<link rel="import" href="../../bower_components/polymer/polymer-element.html">
+<link rel="import" href="../../bower_components/app-storage/app-localstorage/app-localstorage-document.html">
+<link rel="import" href="../../bower_components/iron-ajax/iron-ajax.html">
+<link rel="import" href="../shared-styles.html">
+<link rel="import" href="../redux-store.html">
+
+<dom-module id="weather-element">
+    <template>
+        <style>
+            #temp {
+                text-align: center;
+                font-style: italic;
+            }
+        </style>
+
+        <h3 style="text-align: center">Click the button below to fetch the current temperature in St. Louis, MO!</h3>
+        <div style="display:flex;justify-content:center;align-items:center;">
+            <paper-button id="weatherButton" raised on-click="__fetchWeather">Fetch Weather!</paper-button>
+        </div>
+        <template is="dom-if" if="{{temp}}">
+            <p id="temp">The current temperature in {{city}}{{state}} is {{temp}} degrees {{tempUnit}}! It is {{condition}} today with a wind speed of {{wind}} {{speedUnit}}.</p>
+        </template>
+
+        <iron-ajax
+            id="ajax"
+            method="GET"
+            url="https://query.yahooapis.com/v1/public/yql?q=select%20*%20from%20weather.forecast%20where%20woeid%20in%20(select%20woeid%20from%20geo.places(1)%20where%20text%3D%22st.%20louis%2C%20mo%22)&format=json&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys"
+            handle-as="json"
+            on-response="handleResponse"
```

Comment:
Great job with the iron-ajax usage. However, you shouldn't use both last-response and on-response on the same iron-ajax object. The on-response property actually receives an event that contains the data returned from the service associated with the request. So, on-response is used to transform data after a request is received which is what you seem to be doing in this application. last-response will actually take the data that was returned and, without transforming that data, assign it to the variable specified in the binding. Now, you may wonder why this matters and why this is a big deal and the answer to that question is timing along with code readability/maintainability.

What we want to do is follow a functional programming style in this case. Specifically, we want to reduce what's commonly referred to as side effects from our logic. To do so, we want to avoid storing state and using functions to manipulate that state. To explain why, I'm going to use an example.

Say your code issues a request to a service using iron ajax configured in this way. On response of that request we assume that the following occurs in order: first, the *response* variable is assigned the data returned by the request after which the handleResponse function executes and sets the properties as you've specified. The problem with this is that we can't assume that those operations will occur in that order. So, what happens if the on-response function executes first? What'll actually happen is on-response will execute against the previously assigned *response* value after which *response* will be updated with the new data from the request. This means that a bug was introduced in the code base because your binding won't reflect the new data. Specifically, this.temp, this.tempUnit, etc. 

When it comes to side effects in the code base the reason the functional paradigm is important is because storing state means that other functions that point to that state can actually be operating on data that's "outdated" because that data was stored at some previously unknown time. This scenario is shown in the example above. Passing parameters with the data that concerns you avoids this because the function is executed with all relevant data passed in which means that data must be  "up to date" because you're passing that data into the function which should obviously occur after the new data is received. If this doesn't make sense, please let me know and I'll give a more in-depth explanation on Monday in person. But to show you how I'd rewrite this I'd do the following:

<iron-ajax
id="ajax"
url="..."
handle-as="..."
on-response="handleResponse">
</iron-ajax>
...
handleResponse(e) {
    if (!e.detail || !e.detail.response) return;
    const response = e.detail.response;
    
... To be continued (have to go to dinner :-) )

Reply:
This is extremely helpful thank you!!